lms/services/grouping/service.py

```diff
@@ -203,6 +203,31 @@ def get_groups(
 
         return self._to_groupings(user, groupings, course, self.plugin.group_type)
 
+    def get_known_groupings(self, user, course):
+        """
+        Get the groupings we already know about for the provided user.
+
+        This will include the course, and any sections / groups we have seen
+        before. This does not attempt to retrieve the groups from the LMS.
+
+        This only works for direct children of the course, not children of
+        children.
+        """
+        parent_groups = aliased(Grouping)
+
+        groupings = (
+            # Get groupings which are children of the course
+            self._db.query(Grouping)
+            .filter_by(parent_id=course.id)
+            # Which the user is a part of
+            .join(GroupingMembership)
+            .filter(GroupingMembership.user == user)
+            # Eager load the parent groups
+            .options(subqueryload(parent_groups, Grouping.parent))
+        ).all()
+
+        return [course] + groupings
```

Comment:
Not sure if returning the course + the children is the clearest from the API perspective.

In our database they are the same concept but I could see how that could be confusing for the clients. In `render_lti_context` where we could include "course" as top level concept separate from groups.

And argument against that would be later when actually using the data to query against H where queries would have to put course + groups together to get all relevant annotations.

In Another example the client might need to iterate over groups to find the course using `group.type` to do something particular there. That might be less common that just querying for all the annotations?

Reply:
I think I'm generally in favour of not treating the course as something special. For example, I don't really like the custom model sub-classes and separate service for it, although I know others do. On the whole I think it causes artificial divisions between examples of the same thing.

As for how the service will be used: We have very few service methods that are actually generally used around the app for different purposes. They are almost always used in one specific way. So I think it's probably a good idea to cater to the caller you have rather than a hypothetical caller you imagine you might have in future. If we get a case where we need to filter the two apart it's probably a good time to think again?